components/timers/lib.rs

```diff
@@ -0,0 +1,194 @@
+    for_time: Instant,
+}
+
+impl Ord for ScheduledEvent {
+    fn cmp(&self, other: &ScheduledEvent) -> cmp::Ordering {
+        self.for_time.cmp(&other.for_time).reverse()
+    }
+}
+
+impl PartialOrd for ScheduledEvent {
+    fn partial_cmp(&self, other: &ScheduledEvent) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Eq for ScheduledEvent {}
+impl PartialEq for ScheduledEvent {
+    fn eq(&self, other: &ScheduledEvent) -> bool {
+        std::ptr::eq(self, other)
+    }
+}
+
+/// The message used to send requests to the [`TimerScheduler`] thread.
+enum TimerSchedulerMessage {
+    Request(TimerEventRequest),
+    Timeout,
+}
+
+/// The internal [`TimerScheduler`] which handles incoming requests from the
+/// [`TimerSchedulerProxy`] in a separate thread.
+pub struct TimerScheduler {
+    /// A priority queue of future events, sorted by due time.
+    queue: BinaryHeap<ScheduledEvent>,
+    /// A receiver to receive requests from other threads.
+    receiver: Receiver<TimerSchedulerMessage>,
+}
+
+impl TimerScheduler {
+    /// If there are queued events, wait for a new message or for when the next timer should
+    /// fire. If there are no queued events, just wait for the next event.
+    fn wait_for_event(&self, now: Instant) -> Result<TimerSchedulerMessage, RecvError> {
+        self.queue
+            .peek()
+            .map(
+                |event| match self.receiver.recv_timeout(event.for_time - now) {
+                    Ok(message) => Ok(message),
+                    Err(timeout_error) => match timeout_error {
+                        RecvTimeoutError::Timeout => Ok(TimerSchedulerMessage::Timeout),
+                        RecvTimeoutError::Disconnected => Err(RecvError),
+                    },
+                },
+            )
+            .unwrap_or_else(|| self.receiver.recv())
+    }
+
+    /// Dispatch any timer events from this [`TimerScheduler`]'s `queue` when `now` is
+    /// past the due time of the event.
+    fn dispatch_completed_timers(&mut self, now: Instant) {
+        loop {
+            match self.queue.peek() {
+                // Dispatch the event if its due time is past.
+                Some(event) if event.for_time <= now => {},
+                // Otherwise, we're done dispatching events.
+                _ => break,
+            }
+            // Remove the event from the priority queue (Note this only executes when the
+            // first event has been dispatched
+            self.queue
+                .pop()
+                .expect("Exepcted request")
+                .request
+                .dispatch();
+        }
+    }
+
+    fn schedule_timer(&mut self, request: TimerEventRequest, now: Instant) {
+        let for_time = now + request.duration;
+        self.queue.push(ScheduledEvent { request, for_time });
+    }
+
+    pub fn start() -> TimerSchedulerProxy {
+        let (sender, receiver) = crossbeam_channel::unbounded();
+
+        thread::Builder::new()
```

Comment:
I don't think we need a thread, instead we can integrate this into the script thread, similar to how the previous iteration was integrated into the constellation `select` loop. 

It's hard to spec this precisely here, but I think something along the lines of the below is possible:

1. At the top of [`handle_msgs`](https://github.com/servo/servo/blob/84c59e09b679baadafc30145ac752e37844ac0b8/components/script/script_thread.rs#L1696)
2. Call a variant of `timer_scheduler.wait_for_event` that returns either a crossbream `after`, or `never`. 
3. Add the receiver returned to to the `select!` [here](https://github.com/servo/servo/blob/84c59e09b679baadafc30145ac752e37844ac0b8/components/script/script_thread.rs#L1705)
4. Everytime the `select` wakes-up, so not just in response to the receiver added at 3, run `dispatch_completed_timers`(which will schedule tasks and ensure they are either run in the current iteration, or will wake-up the thread). 

Lastly, I think `TimerSchedulerProxy` can just be `ScriptThread::`, so that we avoid added new concepts(which I think need to be merged into the script-thread logic anyway, as opposed to be independent). 

Please let me know what you think.

Reply:
This sounds good and is a simpler design. I've gone ahead and done this. I have not fully integrated `TimerScheduler` into `ScriptThread`, because it may be that we want to use some kind of timer mechanism in the compositor as well and I was hoping to make this a generic thing that we could reuse later.